sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java

```diff
@@ -0,0 +1,435 @@
+        this.registryImplClient = new AzureContainerRegistryImplBuilder()
+            .url(endpoint)
+            .pipeline(httpPipeline)
+            .apiVersion(version)
+            .buildClient();
+        this.blobsImpl = this.registryImplClient.getContainerRegistryBlobs();
+        this.registriesImpl = this.registryImplClient.getContainerRegistries();
+    }
+
+    /**
+     * This method returns the registry's repository on which operations are being performed.
+     *
+     * @return The name of the repository
+     */
+    public String getRepositoryName() {
+        return this.repositoryName;
+    }
+
+    /**
+     * This method returns the complete registry endpoint.
+     *
+     * @return The registry endpoint including the authority.
+     */
+    public String getEndpoint() {
+        return this.endpoint;
+    }
+
+    /**
+     * Upload the Oci manifest to the repository.
+     * The upload is done as a single operation.
+     * @see <a href="https://github.com/opencontainers/image-spec/blob/main/manifest.md">Oci Manifest Specification</a>
+     * @param manifest The OciManifest that needs to be uploaded.
+     * @return operation result.
+     * @throws ClientAuthenticationException thrown if the client's credentials do not have access to modify the namespace.
+     * @throws NullPointerException thrown if the {@code manifest} is null.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<UploadManifestResult> uploadManifest(OciManifest manifest) {
+        if (manifest == null) {
+            return monoError(logger, new NullPointerException("'manifest' can't be null."));
+        }
+
+        try {
+            byte[] bytes = this.registryImplClient.getSerializerAdapter().serializeToBytes(manifest, SerializerEncoding.JSON);
+            return withContext(context -> this.uploadManifestWithResponse(ByteBuffer.wrap(bytes), context)).flatMap(FluxUtil::toMono);
+        } catch (IOException exception) {
+            return monoError(logger, new UncheckedIOException(exception));
+        }
+    }
+
+    /**
+     * Uploads a manifest to the repository.
+     * The client currently only supports uploading OciManifests to the repository.
+     * And this operation makes the assumption that the data provided is a valid OCI manifest.
+     * <p>
+     * Also, the data is read into memory and then an upload operation is performed as a single operation.
+     * @see <a href="https://github.com/opencontainers/image-spec/blob/main/manifest.md">Oci Manifest Specification</a>
+     * @param data The manifest that needs to be uploaded.
+     * @return operation result.
+     * @throws ClientAuthenticationException thrown if the client's credentials do not have access to modify the namespace.
+     * @throws NullPointerException thrown if the {@code data} is null.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<UploadManifestResult> uploadManifest(BinaryData data) {
```

Comment:
Does the service validate that this data is a valid OCI manifest? Do we get a meaningful error back when an invalid BinaryData is provided?

Reply:
The service supports multiple type of manifests. It does not take manifest type so the error is not like this is not a valid OCI manifest but yes. it gives errors like unknown_manifestType etc. so it is somewhat meaningful.